plugins/eos/gs-plugin-eos.c

```diff
@@ -267,6 +310,330 @@ read_icon_replacement_overrides (GHashTable *replacement_app_lookup)
+				       plugin, NULL);
+}
+
+static void
+os_upgrade_set_download_by_user (GsApp *app, gboolean value)
+{
+	g_autoptr(GVariant) var = g_variant_new_boolean (value);
+	/* we need to assign it to NULL in order to be able to override it
+	 * (safeguard mechanism in GsApp...) */
+	gs_app_set_metadata_variant (app, "eos::DownloadByUser", NULL);
+	gs_app_set_metadata_variant (app, "eos::DownloadByUser", var);
+}
+
+static gboolean
+os_upgrade_get_download_by_user (GsApp *app)
+{
+	GVariant *value = gs_app_get_metadata_variant (app, "eos::DownloadByUser");
+	if (value == NULL)
+		return FALSE;
+	return g_variant_get_boolean (value);
+}
+
+static void
+os_upgrade_force_fetch (EosUpdater *updater)
+{
+	g_auto(GVariantDict) options_dict = G_VARIANT_DICT_INIT (NULL);
+	g_variant_dict_insert (&options_dict, "force", "b", TRUE);
+
+	eos_updater_call_fetch_full (updater,
+				     g_variant_dict_end (&options_dict),
+				     NULL, NULL, NULL);
+}
+
+static gboolean
+os_upgrade_force_fetch_sync (EosUpdater *updater,
+			     GCancellable *cancellable,
+			     GError **error)
+{
+	g_auto(GVariantDict) options_dict = G_VARIANT_DICT_INIT (NULL);
+	g_variant_dict_insert (&options_dict, "force", "b", TRUE);
+
+	return eos_updater_call_fetch_full_sync (updater,
+						 g_variant_dict_end (&options_dict),
+						 cancellable, error);
+}
+
+static void
+app_ensure_installing_state (GsApp *app)
+{
+	/* ensure the state transition to 'installing' is allowed */
+	if (gs_app_get_state (app) != AS_APP_STATE_INSTALLING)
+		gs_app_set_state (app, AS_APP_STATE_AVAILABLE);
+
+	gs_app_set_state (app, AS_APP_STATE_INSTALLING);
+}
+
+static gboolean
+eos_updater_error_is_cancelled (const gchar *error_name)
+{
+	return (g_strcmp0 (error_name, "com.endlessm.Updater.Error.Cancelled") != 0);
+}
+
+static void
+updater_state_changed (GsPlugin *plugin)
+{
+	sync_state_from_updater (plugin);
+}
+
+static void
+updater_version_changed (GsPlugin *plugin)
+{
+	GsApp *app = get_os_upgrade (plugin);
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+	const gchar *version = eos_updater_get_version (priv->updater_proxy);
+
+	gs_app_set_version (app, version);
+}
+
+static void
+disable_os_updater (GsPlugin *plugin)
+{
+	GsPluginData *priv = gs_plugin_get_data (plugin);
+
+	if (priv->upgrade_fake_progress_handler != 0) {
+		g_source_remove (priv->upgrade_fake_progress_handler);
+		priv->upgrade_fake_progress_handler = 0;
+	}
+
+	if (priv->updater_proxy == NULL)
+		return;
+
+	g_signal_handlers_disconnect_by_func (priv->updater_proxy,
+					      G_CALLBACK (updater_state_changed),
+					      plugin);
+	g_signal_handlers_disconnect_by_func (priv->updater_proxy,
+					      G_CALLBACK (updater_version_changed),
+					      plugin);
+
+	g_cancellable_cancel (priv->os_upgrade_cancellable);
+	g_clear_object (&priv->os_upgrade_cancellable);
+
+	g_clear_object (&priv->updater_proxy);
+}
+
+static gboolean
+fake_os_upgrade_progress (GsPlugin *plugin)
```

Comment:
I am not a huge fan of faking progress during installation of files; ideally we could have a different metaphor to use, such as a spinner, in those cases. But I understand why you did it, so we can keep it like this for now; perhaps something good to bring up with upstream though.

Reply:
I agree with you. Let's leave this for now and follow a better solution with upstream as this may happen there too.